crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -1861,6 +1865,87 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn test_block_body_indices() -> eyre::Result<()> {
+        // Initialize random number generator and provider factory
+        let mut rng = generators::rng();
+        let factory = create_test_provider_factory();
+
+        // Generate 10 random blocks and split them into database and in-memory blocks
+        let mut blocks = random_block_range(&mut rng, 0..=10, B256::ZERO, 0..1, None);
+        let (database_blocks, in_memory_blocks) = blocks.split_at_mut(5);
+
+        // Take the first in-memory block and add 7 ommers to it
+        let first_in_mem_block = SealedBlock {
+            ommers: vec![Header::default(); 7],
+            body: vec![
+                random_signed_tx(&mut rng),
+                random_signed_tx(&mut rng),
+                random_signed_tx(&mut rng),
+                random_signed_tx(&mut rng),
+            ],
+            ..in_memory_blocks.first().unwrap().clone()
+        };
+
+        // Insert the first 5 blocks into the database
+        let provider_rw = factory.provider_rw()?;
+        for block in database_blocks.iter_mut() {
+            *block = SealedBlock {
+                body: vec![
+                    random_signed_tx(&mut rng),
+                    random_signed_tx(&mut rng),
+                    random_signed_tx(&mut rng),
+                    random_signed_tx(&mut rng),
+                ],
+                ..block.clone()
+            };
+            provider_rw.insert_historical_block(
+                block.clone().seal_with_senders().expect("failed to seal block with senders"),
+            )?;
+        }
+        provider_rw.commit()?;
+
+        // Create a new provider
+        let provider = BlockchainProvider2::new(factory)?;
+
+        // Insert the first block into the in-memory state
+        let in_memory_block_senders =
+            first_in_mem_block.senders().expect("failed to recover senders");
+        let chain = NewCanonicalChain::Commit {
+            new: vec![ExecutedBlock::new(
+                Arc::new(first_in_mem_block),
+                Arc::new(in_memory_block_senders),
+                Default::default(),
+                Default::default(),
+                Default::default(),
+            )],
+        };
+        provider.canonical_in_memory_state.update_chain(chain);
```

Comment:
can we use `provider_with_random_blocks` here?

Reply:
Here I don't think it's a good idea. Indeed, to properly test the results, I inserted random transactions in the in memory and in database blocks, which is not done in the `provider_with_random_blocks` function. 

When I try to add it to the function directly, it breaks other tests and it's probably not useful elsewhere to add this additional complexity to the util. So I would say that we could stay on this custom made initialization for the test here and that we could eventually see once the tests are all implemented for a more custom and reliable utils test suite, what do you think?

Reply:
makes sense! thanks for the explanation